src/region_binary_search.py

```diff
@@ -0,0 +1,40 @@
+from bed import BedLine
+
+def extract_region(features: list[BedLine],
+                   start: int, end: int) -> list[BedLine]:
+    """Extract region chrom[start:end] and write it to outfile."""
+    index_start = binary_search_region_start(features, start)
+    filtered = list()
+    for i in range(index_start, len(features)):
+        feat = features[i]
+        if feat.chrom_start >= end:
+            break
+        filtered.append(feat)
+    return filtered
+
+def is_overlapping(x: tuple[int,int], y: tuple[int,int]) -> bool:
+    # Credits to Thomas Maillund :)
+    return max(x[0], y[0]) < min(x[1], y[1])
+
+def is_feature_in(feat: BedLine, start: int, end: int):
+    interval = (feat.chrom_start, feat.chrom_end)
+    return is_overlapping(interval, (start, end))
+
+def binary_search_region_start(x: list[BedLine], start: int, low = None, high = None):
+    low, high = 0, len(x)
+    if not x:
+        return 0
```

Comment:
You can avoid this special case if you make the `while` loop test

```python
while low < high:
   ...
return low # return low if you leave the loop
```

That way, if the list is empty, you return index zero, which will not be the first match, but then, if there are no matches, it should give us the first index after where the matches should be, and that zero is indeed.